src/lambda/commands/uploadLambda.ts

```diff
@@ -0,0 +1,246 @@
+
+    return await zipAndUploadDirectory(functionNode, parentDirArr[0].fsPath)
+}
+
+async function runUploadLambdaZipFile(
+    functionNode: LambdaFunctionNode,
+    window = Window.vscode()
+): Promise<telemetry.Result> {
+    const workspaceFolders = vscode.workspace.workspaceFolders || []
+
+    const zipFileArr = await window.showOpenDialog({
+        canSelectFolders: false,
+        canSelectFiles: true,
+        canSelectMany: false,
+        defaultUri: workspaceFolders[0]?.uri,
+        filters: {
+            'ZIP archive': ['zip'],
+        },
+    })
+
+    if (!zipFileArr || zipFileArr.length !== 1) {
+        return 'Cancelled'
+    }
+
+    const zipFile = fs.readFileSync(zipFileArr[0].fsPath)
+
+    return await uploadZipBuffer(functionNode, zipFile)
+}
+
+async function runUploadLambdaWithSamBuild(
+    functionNode: LambdaFunctionNode,
+    window = Window.vscode()
+): Promise<telemetry.Result> {
+    const workspaceFolders = vscode.workspace.workspaceFolders || []
+
+    const parentDirArr = await window.showOpenDialog({
+        canSelectFolders: true,
+        canSelectFiles: false,
+        canSelectMany: false,
+        defaultUri: workspaceFolders[0]?.uri,
+    })
+
+    if (!parentDirArr || parentDirArr.length !== 1) {
+        return 'Cancelled'
+    }
+
+    // TODO: Detect if handler is present and error out prematurely?
+    // TODO: What should we do if no manifest (package.json, requirements.txt, etc.) is present? Directory won't be `sam build`-able.
+
+    try {
+        const invoker = getSamCliContext().invoker
+        const parentDir = parentDirArr[0]
+
+        const tempDir = await makeTemporaryToolkitFolder()
+        ExtensionDisposableFiles.getInstance().addFolder(tempDir)
+        const templatePath = path.join(tempDir, 'template.yaml')
+
+        // TODO: Use an existing template file if it's present?
+
+        await new SamTemplateGenerator()
+            .withFunctionHandler(functionNode.configuration.Handler!)
+            .withResourceName('tempResource')
+            .withRuntime(functionNode.configuration.Runtime!)
+            .withCodeUri(parentDir.fsPath)
+            .generate(templatePath)
+
+        const buildDir = path.join(tempDir, 'output')
+        await new SamCliBuildInvocation({
+            buildDir,
+            templatePath,
+            invoker,
+            skipPullImage: true,
+            useContainer: false,
+            baseDir: parentDir.fsPath,
+        }).execute()
+
+        // TODO: Remove template file if generating one?
+        // TODO: Find correct directory to zip. Needs to be compatible with function's handler since we won't overwrite that.
+
+        return await zipAndUploadDirectory(functionNode, buildDir)
+    } catch (e) {
+        const err = e as Error
+        window.showErrorMessage(err.message)
+        getLogger().error('runUploadLambdaWithSamBuild failed: ', err.message)
+
+        return 'Failed'
+    }
+}
+
+async function zipAndUploadDirectory(
```

Comment:
this could go directly in `runUploadLambdaDirectory`, it's not really abstracting anything except "do two things". 

 `runUploadLambdaDirectory` is already an abstraction over the  action's routine, so breaking it up further into `doXandY` functions is more like indirection than abstraction.

Reply:
This is also used in `runUploadLambdaWithSamBuild` (build dir is zipped and sent here). There's also `uploadZipBuffer` which this function and `runUploadLambdaZipFile` call.